go/oasis-node/cmd/common/metrics/metrics.go

```diff
@@ -179,56 +264,193 @@ func (s *pushService) worker() {
 }
 
 func newPushService() (service.BackgroundService, error) {
-	addr := viper.GetString(cfgMetricsAddr)
-	jobName := viper.GetString(cfgMetricsPushJobName)
-	instanceLabel := viper.GetString(cfgMetricsPushInstanceLabel)
-	interval := viper.GetDuration(cfgMetricsPushInterval)
+	addr := viper.GetString(CfgMetricsAddr)
+	jobName := viper.GetString(CfgMetricsPushJobName)
+	labels := ParseMetricPushLabels(viper.GetString(CfgMetricsPushLabels))
+	interval := viper.GetDuration(CfgMetricsPushInterval)
 
 	if jobName == "" {
 		return nil, fmt.Errorf("metrics: metrics.push.job_name required for push mode")
 	}
-	if instanceLabel == "" {
-		return nil, fmt.Errorf("metrics: metrics.push.instance_label required for push mode")
+	if labels["instance"] == "" {
+		return nil, fmt.Errorf("metrics: at least 'instance' key should be set for metrics.push.labels. Provided labels: %v, viper raw: %v", labels, viper.GetString(CfgMetricsPushLabels))
 	}
 
 	svc := *service.NewBaseBackgroundService("metrics")
 
 	svc.Logger.Debug("Metrics Server Params",
-		"mode", metricsModePush,
+		"mode", MetricsModePush,
 		"addr", addr,
 		"job_name", jobName,
-		"instance_label", instanceLabel,
+		"labels", labels,
 		"push_interval", interval,
 	)
 
+	pusher := push.New(addr, jobName)
+	for k, v := range labels {
+		pusher = pusher.Grouping(k, v)
+	}
+
+	d, err := NewDiskUsageService()
+	if err != nil {
+		return nil, err
+	}
+
 	return &pushService{
 		BaseBackgroundService: svc,
-		pusher:                push.New(addr, jobName).Grouping("instance", instanceLabel),
+		pusher:                pusher,
 		interval:              interval,
+		dService:              d.(*diskService),
 	}, nil
 }
 
 // New constructs a new metrics service.
 func New(ctx context.Context) (service.BackgroundService, error) {
-	mode := viper.GetString(cfgMetricsMode)
+	mode := viper.GetString(CfgMetricsMode)
 	switch strings.ToLower(mode) {
-	case metricsModeNone:
+	case MetricsModeNone:
 		return newStubService()
-	case metricsModePull:
+	case MetricsModePull:
 		return newPullService(ctx)
-	case metricsModePush:
+	case MetricsModePush:
 		return newPushService()
 	default:
 		return nil, fmt.Errorf("metrics: unsupported mode: '%v'", mode)
 	}
 }
 
+type diskService struct {
+	service.BaseBackgroundService
+
+	dataDir string
+	// TODO: Should we monitor I/O of children PIDs as well?
+	pid      int
+	interval time.Duration
+}
+
+func (d *diskService) Start() error {
+	go d.worker()
+	return nil
+}
+
+func (d *diskService) worker() {
+	t := time.NewTicker(d.interval)
+	defer t.Stop()
+
+	var prevRBytes, prevWBytes int
+	for {
+		select {
+		case <-d.Quit():
+			break
+		case <-t.C:
+		}
+
+		{
+			// Compute disk usage of datadir.
+			dataDir := d.dataDir
+			out := bytes.Buffer{}
+			cmd := exec.Command("du", "-sb", dataDir)
```

Comment:
Absolutely not.

https://golang.org/pkg/path/filepath/#Walk
https://golang.org/pkg/os/#Stat